### PR TITLE
Fix redundant resolves in comms vat interaction

### PR DESCRIPTION
**Pull request by FUDCo**
This appears to be the fix for #2393 

The issue was nothing like we imagined.

Turns out the comms vat already had some code that was kinda sorta doing the "fill in reintroduced promises" thing that our new multi-promise resolve does, sorta kinda.  This is a simple change, but it probably warrants more scrutiny than most, just because.

### CODE REVIEW

**Comment by FUDCo** [comment]
I didn't expect lint to trip over this change but it does.  It turns out that the change renders `resolveToRemote` inside `clist-outbound.js` nugatory.  I will clean that up with a separate commit; since it makes for a much more invasive change it will be nice to isolate it.

File: packages/SwingSet/src/controller.js [hunk 1]
```diff
@@ -68,9 +72,17 @@ export async function makeSwingsetController(
   harden(console);
 
   // FIXME: Put this somewhere better.
-  process.on('unhandledRejection', e =>
-    console.error('UnhandledPromiseRejectionWarning:', e),
-  );
+  const handlers = process.listeners('unhandledRejection');
```
Comment (warner) [inline packages/SwingSet/src/controller.js]: This is good, but let's put it in a separate PR.

File: packages/SwingSet/src/vats/comms/clist-outbound.js [hunk 2]
```diff
@@ -77,14 +71,8 @@ export function makeOutbound(state, stateKit) {
...
 
-    if (p.resolved) {
-      // we must send the resolution *after* the message which introduces it
-      const { remoteID } = remote;
-      Promise.resolve().then(() =>
-        resolveToRemote(remoteID, [[vpid, p.rejected, p.data]]),
-      );
-    } else {
-      // or arrange to send it later, once it resolves
+    if (!p.resolved) {
```
Comment (warner) [inline packages/SwingSet/src/vats/comms/clist-outbound.js:74]: Will this new version correctly handle the case where a previously-resolved Promise is included in the arguments of a `send`? I can see how the `resolve` case will work better without this extra `resolveToRemote()`, but I suspect the `send` case really needs it.

We talked a while ago about moving this sort of "send extra messages" code out of the translation function and into a post-translation clause that looks at the `slots`, figures out which ones are Promises, figures out which of *those* are resolved and new, and then sends the necessary `resolve` messages in the same turn as the `send`. I think I'm liking that option more and more.

**Review by warner — CHANGES_REQUESTED** [review]
We need to split the `controller.js` into a separate PR, and then either add a test or identify an existing test for the resolved-promise-as-argument-of-send question, probably in message-patterns.js.

**Comment by warner** [comment]
That test might be tricky. The kernel will never send a `send` into a vat whose arguments reference an already-resolved promise, because it would have retired that `vpid` already. So I think we need to provoke this situation with a three-party forwarding: commsA receives a `send` from commsB that targets an object from commsC, and its arguments reference a promise that commsA knows is resolved. CommsA will have sent a message to commsB about the resolution, but the `send` is plausibly generated before commsB received it, so it's legitimate for commsB to have been ignorant about its resolution.

File: packages/SwingSet/src/vats/comms/clist-outbound.js (hunk 2 above)
Comment (FUDCo) [reply packages/SwingSet/src/vats/comms/clist-outbound.js:74]: This makes sense.  The kernel does that, and we're finding a lot of the comms vat logic mirrors the kernel logic.

File: packages/SwingSet/src/controller.js (hunk 1 above)
Comment (FUDCo) [reply packages/SwingSet/src/controller.js]: done

**Comment by FUDCo** [comment]
After soaking in the problem space for a few days, I think I've convinced myself that the potential problem @warner raises concerns about above won't really happen (or, rather, there might be some edge cases where it could happen, but if so it would be benign).

My reasoning is this: a message send necessarily originates in an end point vat (i.e., not a comms vat), so whenever the sender has to follow its `send` syscall with a trailing `resolve` syscall to communicate the resolution of resolved promises that were introduced by the send itself, it follows that any and all such promises both originated in and were decided by the sending vat, and not (to be a bit redundant for emphasis) any comms vats that handle the message on its way to its ultimate destination.  This sequence of a message send followed by a resolve translates into a `send` followed by a `notify` being delivered into the sending SwingSet's comms vat as separate events in separate cranks.  Thus the `notify` is temporally disconnected from the preceding `send` it is associated with -- there's nothing that directly indicates to the comms vat that the promises in the `notify` have anything to do with some `send` that passed through its hands earlier.  From the comms vat's perspective, it's just like any other `notify`, and so should follow the same logic as any other `notify`.  However, the promises that were introduced by the `send` are only relevant in the context of the message being sent -- that's the only place they're referenced -- so their resolutions are only of interest to the receiver of that message, and not other unrelated parties who might happen to be reachable via the comms vat.  This is different from the case of a naked `notify`, where the primary promise being notified could have multiple remote subscribers visible to the comms vat.

If you accept this reasoning, then it should be safe to land this PR and close out the issue that it addresses.

**Comment by FUDCo** [comment]
After some discussion, @warner and I have concluded that the best path forward is to land this change and file a new ticket for the issue flagged in his comment above, as a precursor to a larger refactoring that we've identified (ticket RSN) that should come between this change and the fixes that will be needed for the aforementioned problem.

**Review by warner — APPROVED** [review]
Agreed. We think this will expose a different bug than the one it fixes, and that we don't yet have test coverage for that bug. But we know what that test needs to do. I'll write up a ticket describing it, and it should get fixed by the "collector" / "send `resolve` after `send`" changes that @FUDCo has begun, but which will be easier to implement after we refactor the comms vat to have a real kernel-facing clist.
